sd_bandits/obp_extensions/policy.py

```diff
@@ -85,3 +88,130 @@ def update_params(self, action: int, reward: float) -> None:
+        segment_policy_pairs = [(i, deepcopy(self.base_policy)) for i in range(self.n_segments)]
+        self.segment_policies = dict(segment_policy_pairs)
+        self.update_dict = dict([(i, []) for i in range(self.n_segments)])
+        
+    @property
+    def action_counts(self):
+        """Returns action counts"""
+        action_counts = np.zeros(self.n_actions)
+        for i in range(self.n_segments):
+            action_counts += self.segment_policies[i].action_counts
+        return action_counts
+    
+    @property
+    def reward_counts(self):
+        """Returns reward counts"""
+        reward_counts = np.zeros(self.n_actions)
+        for i in range(self.n_segments):
+            reward_counts += self.segment_policies[i].action_counts * self.segment_policies[i].reward_counts
+        reward_counts = np.divide(reward_counts, self.action_counts, where=self.action_counts != 0)
+        return reward_counts
+        
+        
+    def select_action(self, segment: int) -> np.ndarray:
+        """Select a list of actions.
+        
+        Parameters
+        segment: int
+            The user's segment
+        
+        Returns
+        ----------
+        selected_actions: array-like, shape (len_list, )
+            List of selected actions.
+
+        """
+        
+        return self.segment_policies[segment].select_action()
+
+
+    def update_params(self, action: int, reward: float, segment: int) -> None:
+        """Update policy parameters.
+        Same as obp.policy.EpsilonGreedy
+
+        Parameters
+        ----------
+        action: int
+            Selected action by the policy.
+
+        reward: float
+            Observed reward for the chosen action and position.
+        
+        segment: int
+            The user's segment
+
+        """
+        # first update n_trial for the policy
+        self.n_trial += 1
+        
+        # then add action and reward to the update dictionary
+        self.update_dict[segment].append((action, reward))
+
+        # if the policy should update its parameters, call _update_all_params
+        if self.n_trial % self.batch_size == 0:
+            self._update_all_params()
+                    
+    def _update_all_params(self) -> None:
+        """Updates all segment policy parameters.
+        Only triggered when n_trials mod batch_size == 0
+        """
+        for seg_, action_reward_list in self.update_dict.items():
+            # setting each segment policy's batch_size to the number of times
+            # it has to update. so it will trigger a full at the end of the 
+            # inner for-loop
+            self.segment_policies[seg_].batch_size =  len(action_reward_list)
+            self.segment_policies[seg_].n_trial = 0
+            for (action_, reward_) in action_reward_list:
+                self.segment_policies[seg_].update_params(action_, reward_)
+                
+            # then reset the update dict
+            self.update_dict[seg_] = []
```

Comment:
i think, rather than mutating a dictionary while iterating over it, it's safer to just reset the whole dictionary when the loop is done to `dict([(i, []) for i in range(self.n_segments)])`

Reply:
Ah that's what I had first lol. Changing now